src/ML-UF3/pair_uf3.cpp

```diff
@@ -0,0 +1,1313 @@
+    for (int j = i; j < num_of_elements + 1; j++) {
+      UFBS2b[i][j] = uf3_pair_bspline(lmp, n2b_knot[i][j], n2b_coeff[i][j],
+              knot_spacing_type_2b[i][j]);
+      UFBS2b[j][i] = UFBS2b[i][j];
+    }
+    if (pot_3b) {
+      for (int j = 1; j < num_of_elements + 1; j++) {
+        for (int k = j; k < num_of_elements + 1; k++) {
+          std::string key = std::to_string(i) + std::to_string(j) + std::to_string(k);
+          UFBS3b[i][j][k] =
+              uf3_triplet_bspline(lmp, n3b_knot_matrix[i][j][k], n3b_coeff_matrix[key],
+                      knot_spacing_type_3b[i][j][k]);
+          std::string key2 = std::to_string(i) + std::to_string(k) + std::to_string(j);
+          UFBS3b[i][k][j] =
+              uf3_triplet_bspline(lmp, n3b_knot_matrix[i][k][j], n3b_coeff_matrix[key2],
+                      knot_spacing_type_3b[i][k][j]);
+        }
+      }
+    }
+  }
+}
+
+void PairUF3::compute(int eflag, int vflag)
+{
+  int i, j, k, ii, jj, kk, inum, jnum, itype, jtype, ktype;
+  double xtmp, ytmp, ztmp, delx, dely, delz, evdwl, fpair, fx, fy, fz;
+  double del_rji[3], del_rki[3], del_rkj[3];
+  double fij[3], fik[3], fjk[3];
+  double fji[3], fki[3], fkj[3];
+  double Fi[3], Fj[3], Fk[3];
+  double rsq, rij, rik, rjk;
+  int *ilist, *jlist, *numneigh, **firstneigh;
+
+  ev_init(eflag, vflag);
+
+  double **x = atom->x;
+  double **f = atom->f;
+  int *type = atom->type;
+  int nlocal = atom->nlocal;
+  int newton_pair = force->newton_pair;
+
+  inum = list->inum;
+  ilist = list->ilist;
+  numneigh = list->numneigh;
+  firstneigh = list->firstneigh;
+  // loop over neighbors of my atoms
+  for (ii = 0; ii < inum; ii++) {
+    evdwl = 0;
+    i = ilist[ii];
+    xtmp = x[i][0];
+    ytmp = x[i][1];
+    ztmp = x[i][2];
+    itype = type[i];
+    jlist = firstneigh[i];
+    jnum = numneigh[i];
+    int numshort = 0;
+    for (jj = 0; jj < jnum; jj++) {
+      fx = 0;
+      fy = 0;
+      fz = 0;
+      j = jlist[jj];
+      j &= NEIGHMASK;
+
+      delx = xtmp - x[j][0];
+      dely = ytmp - x[j][1];
+      delz = ztmp - x[j][2];
+
+      rsq = delx * delx + dely * dely + delz * delz;
+      jtype = type[j];
+      if (rsq < cutsq[itype][jtype]) {
+        rij = sqrt(rsq);
+
+        if (pot_3b) {
+          if (rij <= cut_3b_list[itype][jtype]) {
+            neighshort[numshort] = j;
+            if (numshort >= maxshort - 1) {
+              maxshort += maxshort / 2;
+              memory->grow(neighshort, maxshort, "pair:neighshort");
+            }
+            numshort = numshort + 1;
+          }
+        }
+
+        double *pair_eval = UFBS2b[itype][jtype].eval(rij);
+
+        fpair = -1 * pair_eval[1] / rij;
+
+        fx = delx * fpair;
+        fy = dely * fpair;
+        fz = delz * fpair;
+
+        f[i][0] += fx;
+        f[i][1] += fy;
+        f[i][2] += fz;
+        f[j][0] -= fx;
+        f[j][1] -= fy;
+        f[j][2] -= fz;
+
+        if (eflag) evdwl = pair_eval[0];
```

Comment:
Should I also pass eflag to the eval function? Something like `UFBS2b[itype][jtype].eval(rij, evdwl, eflag)`? This will save some computations as energy evaluation is not needed on every timestep?

Reply:
If the energy calculation is costly enough, then yes, you can pass eflag and let `eval` save some time. But you would know better whether that's worth it.

As a style thing I'd prefer to keep the reference doubles together after the value doubles:

```
double CLASSNAME::eval(
  double rij,
  int eflag,
  double& evdwl // return energy by ref
){ ...
```